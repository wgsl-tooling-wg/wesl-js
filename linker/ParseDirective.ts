import {
  anyThrough,
  kind,
  opt,
  or,
  repeat,
  req,
  seq,
  setTraceNames,
  tokens,
  tracing,
} from "@wesl/mini-parse";
import { gleamImport } from "./GleamImport.ts";
import {
  argsTokens,
  moduleTokens,
} from "./MatchWgslD.ts";
import { eolf, makeElem } from "./ParseSupport.ts";

/* parse #directive enhancements to wgsl: #export, etc. */

const argsWord = kind(argsTokens.arg);
const fromWord = or(argsWord, kind(argsTokens.relPath));

const fromClause = seq(
  "from",
  or(fromWord.tag("from"), seq('"', fromWord.tag("from"), '"')),
);

/** #export <foo> <(a,b)> EOL */
export const exportDirective = seq(or("#export", "export"), opt(eolf)).map(
  r => {
    const e = makeElem("export", r, ["args"]);
    r.app.state.push(e);
  },
);

const moduleDirective = seq(
  or("module", "#module"),
  tokens(moduleTokens, req(kind(moduleTokens.moduleName).tag("name"))),
  eolf,
).map(r => {
  const e = makeElem("module", r);
  e.name = normalizeModulePath(r.tags.name[0]);
  r.app.state.push(e);
});

function normalizeModulePath(name: string): string {
  if (name.includes("::")) {
    const result = name.split("::").join("/");
    return result;
  }
  return name;
}

export const directive = tokens(
  argsTokens,
  seq(repeat("\n"), or(exportDirective, gleamImport, moduleDirective)),
);

if (tracing) {
  setTraceNames({
    fromClause,
    exportDirective,
    moduleDirective,
    directive,
  });
}
