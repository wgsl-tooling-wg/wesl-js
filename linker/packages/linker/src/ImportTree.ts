export class ImportTree {
  /** segments in path order */
  constructor(public segments: PathSegment[]) {}
}

export type PathSegment = SimpleSegment | ImportTree | SegmentList;

export class SimpleSegment {
  constructor(
    public name: string,
    public as?: string,
    public args?: string[], // generic args (only allowed on final segment). TODO drop
  ) {}
}

/** or choices for this path segment */
export class SegmentList {
  constructor(public list: PathSegment[]) {}
}

export function treeToString(tree: ImportTree): string {
  return tree.segments.map(s => segmentToString(s)).join("/");
}

function segmentToString(segment: PathSegment): string {
  if (segment instanceof SimpleSegment) {
    const { name, as, args } = segment;
    const asMsg = as ? ` as ${as}` : "";
    const argsMsg = args ? `(${args.join(", ")})` : "";
    return `${name}${argsMsg}${asMsg}`;
  }
  if (segment instanceof SegmentList) {
    return `{${segment.list.map(s => segmentToString(s)).join(", ")}}`;
  }
  if (segment instanceof ImportTree) {
    return `(${treeToString(segment)})`;
  }
  return `|unknown segment type ${(segment as any).constructor.name}|`;
}
