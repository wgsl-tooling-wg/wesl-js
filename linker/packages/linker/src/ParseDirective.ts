import {
  anyThrough,
  kind,
  opt,
  or,
  repeat,
  req,
  seq,
  setTraceNames,
  tokens,
  tracing,
} from "mini-parse";
import { gleamImport } from "./GleamImport.js";
import { argsTokens, lineCommentTokens, mainTokens } from "./MatchWgslD.js";
import { eolf } from "./ParseSupport.js";

/* parse #directive enhancements to wgsl: #export, etc. */

const argsWord = kind(argsTokens.arg);
const fromWord = or(argsWord, kind(argsTokens.relPath));

const fromClause = seq("from", or(fromWord, seq('"', fromWord, '"')));

/** #export <foo> <(a,b)> EOL */
export const exportDirective = seq(or("#export", "export"), opt(eolf));

export const directive = tokens(
  argsTokens,
  seq(repeat("\n"), or(exportDirective, gleamImport)),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

if (tracing) {
  setTraceNames({
    fromClause,
    exportDirective,
    skipToEol,
    lineComment,
    directive,
  });
}
