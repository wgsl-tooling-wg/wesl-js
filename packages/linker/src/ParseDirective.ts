import {
  anyThrough,
  kind,
  opt,
  or,
  repeat,
  req,
  seq,
  setTraceNames,
  tokens,
  tracing,
} from "mini-parse";
import { gleamImport } from "./GleamImport.js";
import {
  argsTokens,
  lineCommentTokens,
  mainTokens,
  moduleTokens,
} from "./MatchWgslD.js";
import { eolf, makeElem } from "./ParseSupport.js";

/* parse #directive enhancements to wgsl: #export, etc. */

const argsWord = kind(argsTokens.arg);
const fromWord = or(argsWord, kind(argsTokens.relPath));

const fromClause = seq(
  "from",
  or(fromWord.tag("from"), seq('"', fromWord.tag("from"), '"')),
);

/** #export <foo> <(a,b)> EOL */
export const exportDirective = seq(or("#export", "export"), opt(eolf)).map(
  r => {
    const e = makeElem("export", r, ["args"]);
    r.app.state.push(e);
  },
);

const moduleDirective = seq(
  or("module", "#module"),
  tokens(moduleTokens, req(kind(moduleTokens.moduleName).tag("name"))),
  eolf,
).map(r => {
  const e = makeElem("module", r);
  e.name = normalizeModulePath(r.tags.name[0]);
  r.app.state.push(e);
});

function normalizeModulePath(name: string): string {
  if (name.includes("::")) {
    const result = name.split("::").join("/");
    return result;
  }
  return name;
}

export const directive = tokens(
  argsTokens,
  seq(repeat("\n"), or(exportDirective, moduleDirective)),
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment */
export const lineComment = seq(tokens(mainTokens, "//"), skipToEol);

if (tracing) {
  setTraceNames({
    fromClause,
    exportDirective,
    skipToEol,
    lineComment,
    moduleDirective,
    directive,
  });
}
