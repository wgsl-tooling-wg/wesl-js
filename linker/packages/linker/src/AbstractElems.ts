import { ImportTree } from "./ImportTree.ts";
import { DeclIdent, RefIdent, SrcModule } from "./Scope.ts";

/** 
 * Structures to describe the 'interesting' parts of a WESL source file.
 *
 * The parts of the source that need to analyze further in the linker 
 * are pulled out into these structures.
 * 
 * The parts that are uninteresting the the linker are recorded
 * as 'TextElem' nodes, which are generally just copied to the output WGSL
 * along with their containing element.
 */
export type AbstractElem =
  | AliasElem
  | AttributeElem
  | ConstElem
  | ExpressionElem
  | ImportElem
  | ConstAssertElem
  | FnElem
  | RefIdentElem
  | DeclIdentElem
  | ModuleElem
  | NameElem
  | OverrideElem
  | ParamElem
  | StructElem
  | StructMemberElem
  | TextElem
  | TypeRefElem
  | GlobalVarElem
  | VarElem;

export type DeclarationElem =
  | AliasElem
  | ConstElem
  | OverrideElem
  | ParamElem
  | FnElem
  | StructElem
  | GlobalVarElem
  | VarElem;

export interface AbstractElemBase {
  kind: string;
  start: number;
  end: number;
}

export interface ElemWithContents extends AbstractElemBase {
  contents: AbstractElem[];
}

/** a raw bit of text in WESL source that's typically copied to the linked WGSL. 
 e.g. a keyword  like 'var' 
 or a phrase we needn't analyze further like '@diagnostic(off,derivative_uniformity)'
*/
export interface TextElem extends AbstractElemBase {
  kind: "text";
  srcModule: SrcModule;
}

export interface ImportElem extends ElemWithContents {
  kind: "import";
  imports: ImportTree;
}

export type TypeTemplateParameter = TypeRefElem | ExpressionElem | string;

/** an expression (generally we don't need details of expressions, just their contained idents */
export interface ExpressionElem extends ElemWithContents {
  kind: "expression";
}

/** a reference to a type, like 'f32', or 'MyStruct', or 'ptr<storage, array<f32>, read_only>'   */
export interface TypeRefElem extends ElemWithContents {
  kind: "type";
  name: RefIdent | string;
  templateParams?: TypeTemplateParameter[];
}

/** an identifier that refers to a declaration */
export interface RefIdentElem extends AbstractElemBase {
  kind: RefIdent["kind"];
  ident: RefIdent;
  srcModule: SrcModule;
}

/** a declaration identifier */
export interface DeclIdentElem extends AbstractElemBase {
  kind: DeclIdent["kind"];
  ident: DeclIdent;
  srcModule: SrcModule;
}

/** a parameter in a function declaration */
export interface ParamElem extends ElemWithContents {
  kind: "param";
  name: DeclIdentElem;
  typeRef: TypeRefElem;
}

/** an attribute like '@compute' or '@binding(0)' */
export interface AttributeElem extends ElemWithContents {
  kind: "attribute";
  name: string;
  params?: ExpressionElem[];
}

/** a variable declaration */
export interface VarElem extends ElemWithContents {
  kind: "var";
  name: DeclIdentElem;
  typeRef?: TypeRefElem;
}

/** a global variable declaration (at the root level) */
export interface GlobalVarElem extends ElemWithContents {
  kind: "gvar";
  name: DeclIdentElem;
  typeRef?: TypeRefElem;
}

/** a const declaration */
export interface ConstElem extends ElemWithContents {
  kind: "const";
  name: DeclIdentElem;
  typeRef?: TypeRefElem;
}

/** an override declaration */
export interface OverrideElem extends ElemWithContents {
  kind: "override";
  name: DeclIdentElem;
  typeRef?: TypeRefElem;
}

/** an entire file */
export interface ModuleElem extends ElemWithContents {
  kind: "module";
}

/** an alias statement */
export interface AliasElem extends ElemWithContents {
  kind: "alias";
  name: DeclIdentElem;
  typeRef: TypeRefElem;
}

/** a const_assert statement */
export interface ConstAssertElem extends ElemWithContents {
  kind: "assert";
}

/** a struct declaration */
export interface StructElem extends ElemWithContents {
  kind: "struct";
  name: DeclIdentElem;
  members: StructMemberElem[];
  bindingStruct?: true; // used later during binding struct transformation
}

/** a member of a struct declaration */
export interface StructMemberElem extends ElemWithContents {
  kind: "member";
  name: NameElem;
  attributes?: AttributeElem[];  
  typeRef: TypeRefElem;
}

/** a name (e.g. a struct member name) that doesn't need to be an Ident */
export interface NameElem extends AbstractElemBase {
  kind: "name";
  name: string;
  srcModule: SrcModule;
}

/** a function declaration */
export interface FnElem extends ElemWithContents {
  kind: "fn";
  name: DeclIdentElem;
  params: ParamElem[];
  returnType?: TypeRefElem;
}
