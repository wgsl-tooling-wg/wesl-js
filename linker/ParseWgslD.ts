import {
  anyNot,
  anyThrough,
  eof,
  ExtendedResult,
  kind,
  matchingLexer,
  opt,
  or,
  Parser,
  ParserContext,
  ParserInit,
  preParse,
  repeat,
  repeatPlus,
  req,
  seq,
  setTraceName,
  simpleParser,
  SrcMap,
  tokens,
  tracing,
  withSep,
} from "@wesl/mini-parse";
import { AbstractElem, TypeNameElem, TypeRefElem } from "./AbstractElems.ts";
import { identTokens, mainTokens } from "./MatchWgslD.ts";
import { directive } from "./ParseDirective.ts";
import {
  comment,
  literal,
  makeElem,
  op,
  unknown,
  word,
} from "./ParseSupport.ts";

/** parser that recognizes key parts of WGSL and also directives like #import */

const longIdent = kind(identTokens.longIdent);

// prettier gets confused if we leave the quoted parens inline so make consts for them here
const lParen = "(";
const rParen = ")";

export interface ParseState {
  params: Record<string, any>; // user provided params to templates, code gen and #if directives
}

// TODO: Check the following
// - translation_unit
// - global_decl
// - global_value_decl
// - global_directive
// - diagnostic_rule_name
// - diagnostic_control

const attribute = seq(
  "@",
  req(or(
    // These attributes have no arguments
    or("compute", "const", "fragment", "invariant", "must_use", "vertex"),
    // These attributes have arguments, but the argument doesn't have any identifiers
    seq(
      or("interpolate", "builtin", "diagnostic"),
      req(() => argument_expression_list), // TODO: Throw away the identifiers
    ),
    // These are normal attributes
    seq(
      or(
        "workgroup_size",
        "align",
        "binding",
        "blend_src",
        "group",
        "id",
        "location",
        "size",
      ),
      req(() => argument_expression_list),
    ),
    // Everything else is also a normal attribute, it might have an expression list
    seq(kind(mainTokens.word), opt(() => argument_expression_list)),
  )),
);

const argument_expression_list = seq(
  "(",
  withSep(",", () => expression),
  req(")"),
);

const opt_attributes = repeat(attribute);
const possibleTypeRef = Symbol("typeRef");

const globalDirectiveOrAssert = seq(
  or("diagnostic", "enable", "requires", "const_assert"),
  req(anyThrough(";")),
).map((r) => {
  const e = makeElem("globalDirective", r);
  r.app.state.push(e);
});

/** parse an identifier into a TypeNameElem */
export const typeNameDecl = req(word.tag("name")).map((r) => {
  return makeElem("typeName", r, ["name"]) as TypeNameElem; // fix?
});

/** parse an identifier into a TypeNameElem */
export const fnNameDecl = req(word.tag("name"), "missing fn name").map((r) => {
  return makeElem("fnName", r, ["name"]);
});

/** find possible references to user structs in this type specifier and any templates */
export const type_specifier: Parser<TypeRefElem[]> = seq(
  tokens(identTokens, longIdent.tag(possibleTypeRef)),
  () => opt_template_args,
).map((r) =>
  r.tags[possibleTypeRef].map((name) => {
    const e = makeElem("typeRef", r as ExtendedResult<any>);
    e.name = name;
    return e as Required<typeof e>;
  })
);

const optionally_typed_ident = seq(
  word,
  opt(seq(":", type_specifier.tag("typeRefs"))),
);

export const structMember = seq(
  opt_attributes,
  word.tag("name"),
  ":",
  req(type_specifier.tag("typeRefs")),
).map((r) => {
  return makeElem("member", r, ["name", "typeRefs"]);
});

export const structDecl = seq(
  "struct",
  req(typeNameDecl).tag("nameElem"),
  req("{"),
  withSep(",", structMember).tag("members"),
  req("}"),
).map((r) => {
  const e = makeElem("struct", r, ["members"]);
  const nameElem = r.tags.nameElem[0];
  e.nameElem = nameElem;
  e.name = nameElem.name;
  r.app.state.push(e);
});

/** Also covers func_call_statement.post.ident */
export const fn_call = seq(
  longIdent
    .tag("name")
    .map((r) => makeElem("call", r, ["name"]))
    .tag("calls"), // we collect this in fnDecl, to attach to FnElem
  () => opt_template_args,
  argument_expression_list,
);

// prettier-ignore
const fnParam = seq(
  opt_attributes,
  word,
  opt(seq(":", req(type_specifier.tag("typeRefs")))),
);

const fnParamList = seq(lParen, withSep(",", fnParam), rParen);

/** Covers variable_decl and the 'var' case in global_decl */
const variable_decl = seq(
  "var",
  () => opt_template_args,
  optionally_typed_ident,
  opt(seq("=", () => expression)),
);

/** Aka template_elaborated_ident.post.ident */
const opt_template_args = opt(
  seq(
    "<",
    withSep(",", () => template_arg_expression, {
      requireOne: true,
    }),
    ">",
  ).tag("template"),
);

const primary_expression = or(
  literal,
  seq(
    word.tag("ident"),
    opt_template_args,
    opt(
      argument_expression_list,
    ),
  ),
  seq("(", () => expression, req(")")),
);
const component_or_swizzle = repeatPlus(
  or(
    seq(".", word),
    seq("[", () => expression, req("]")),
  ),
);

/**
 * bitwise_expression.post.unary_expression
 * & ^ |
 * expression
 * && ||
 * relational_expression.post.unary_expression
 * > >= < <= != ==
 * shift_expression.post.unary_expression
 * % * / + - << >>
 */
const makeExpressionOperator = (isTemplate: boolean) => {
  const allowedOps = (
    "& | ^ << <= < != == % * / + -" + (isTemplate ? "" : " && || >> >= >")
  ).split(" ").map(op);
  return or(...allowedOps)
    .traceName("operator")
    .trace({
      shallow: true,
    });
};
const unary_expression: Parser<any> = or(
  seq(
    or(..."! & * - ~".split(" "))
      .traceName("unary_op")
      .trace({
        shallow: true,
      }),
    () => unary_expression,
  ),
  seq(primary_expression, opt(component_or_swizzle)),
);
const makeExpression = (isTemplate: boolean) => {
  return seq(
    unary_expression,
    repeat(seq(makeExpressionOperator(isTemplate), unary_expression)),
  );
};

export const expression = makeExpression(false);
const template_arg_expression = makeExpression(true);

const compound_statement = seq(
  opt_attributes,
  "{",
  repeat(() => statement),
  "}",
);

const for_init = or(
  fn_call,
  () => variable_or_value_statement,
  () => variable_updating_statement,
);

const for_update = or(
  fn_call,
  () => variable_updating_statement,
);

const for_statement = seq(
  opt_attributes,
  "for",
  req(seq("(", opt(for_init), ";", opt(expression), ";", opt(for_update), ")")),
);
const if_statement = seq(
  opt_attributes,
  "if",
  req(seq(expression, compound_statement)),
  repeat(
    seq("else", "if", req(seq(expression, compound_statement))),
  ),
  opt(seq("else", req(compound_statement))),
);
const loop_statement = seq(
  opt_attributes,
  "loop",
  opt_attributes,
  req(seq(
    "{",
    repeat(() => statement),
    opt(
      seq(
        "continuing",
        opt_attributes,
        "{",
        repeat(() => statement),
        opt(seq("break", "if", expression, ";")),
        "}",
      ),
    ),
    "}",
  )),
);

const case_selector = or("default", expression);
const switch_clause = or(
  seq(
    "case",
    withSep(",", case_selector, { requireOne: true }),
    opt(":"),
    compound_statement,
  ),
  seq("default", opt(":"), compound_statement),
);

const switch_body = seq(opt_attributes, "{", repeatPlus(switch_clause), "}");
const switch_statement = seq(opt_attributes, "switch", expression, switch_body);

const while_statement = seq(
  opt_attributes,
  "while",
  expression,
  compound_statement,
);

const statement: Parser<any> = or(
  for_statement,
  if_statement,
  loop_statement,
  switch_statement,
  while_statement,
  compound_statement,
  seq("break", ";"),
  seq("continue", ";"),
  seq(";"),
  seq("const_assert", expression, ";"),
  seq("discard", ";"),
  seq("return", opt(expression), ";"),
  seq(fn_call, ";"),
  seq(() => variable_or_value_statement, ";"),
  seq(() => variable_updating_statement, ";"),
);

const lhs_expression: Parser<any> = or(
  seq(
    word.tag("ident"),
    opt(component_or_swizzle),
  ),
  seq("(", () => lhs_expression, ")", opt(component_or_swizzle)),
  seq("&", () => lhs_expression),
  seq("*", () => lhs_expression),
);

const variable_or_value_statement = or(
  // Also covers the = expression case
  variable_decl,
  seq("const", optionally_typed_ident, "=", expression),
  seq("let", optionally_typed_ident, "=", expression),
);
const variable_updating_statement = or(
  seq(
    lhs_expression,
    or(
      "=",
      op("<<="),
      op(">>="),
      op("%="),
      op("&="),
      op("*="),
      op("+="),
      op("-="),
      op("/="),
      op("^="),
      op("|="),
    ),
    expression,
  ),
  seq(lhs_expression, or(op("++"), op("--"))),
  seq("_", "=", expression),
);

export const fn_decl = seq(
  opt_attributes,
  "fn",
  req(fnNameDecl).tag("nameElem"),
  req(fnParamList),
  opt(seq("->", opt_attributes, type_specifier.tag("typeRefs"))),
  req(compound_statement),
).trace().map((r) => {
  const e = makeElem("fn", r);
  const nameElem = r.tags.nameElem[0];
  e.nameElem = nameElem as Required<typeof nameElem>;
  e.name = nameElem.name;
  e.calls = r.tags.calls || [];
  e.typeRefs = r.tags.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalVar = seq(
  opt_attributes,
  or("const", "override", "var"),
  opt_template_args,
  word.tag("name"),
  opt(seq(":", req(type_specifier.tag("typeRefs")))),
  req(anyThrough(";")),
).map((r) => {
  const e = makeElem("var", r, ["name"]);
  e.typeRefs = r.tags.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalAlias = seq(
  "alias",
  req(word.tag("name")),
  req("="),
  req(type_specifier).tag("typeRefs"),
  req(";"),
).map((r) => {
  const e = makeElem("alias", r, ["name", "typeRefs"]);
  r.app.state.push(e);
});

const globalDecl = or(fn_decl, globalVar, globalAlias, structDecl, ";");

const rootDecl = or(globalDirectiveOrAssert, globalDecl, directive, unknown);

const root = preParse(comment, seq(repeat(rootDecl), eof()));

export function parseWgslD(
  src: string,
  srcMap?: SrcMap,
  params: Record<string, any> = {},
  maxParseCount: number | undefined = undefined,
  grammar = root,
): AbstractElem[] {
  const lexer = matchingLexer(src, mainTokens);
  const state: AbstractElem[] = [];
  const context: ParseState = { params };
  const app = {
    context,
    state,
  };
  const init: ParserInit = {
    lexer,
    app,
    srcMap,
    maxParseCount,
  };

  grammar.parse(init);

  return app.state;
}

if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    globalDirectiveOrAssert,
    type_specifier,
    structMember,
    structDecl,
    fn_call,
    fnParam,
    fnParamList,
    opt_template_args,
    primary_expression,
    component_or_swizzle,
    expression,
    statement,
    compound_statement,
    case_selector,
    switch_clause,
    switch_body,
    switch_statement,
    fn_decl,
    globalVar,
    globalAlias,
    globalDecl,
    rootDecl,
    root,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
