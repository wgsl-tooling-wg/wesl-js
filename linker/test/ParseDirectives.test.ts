import { expect, test } from "vitest";
import { ModuleElem, TreeImportElem } from "../AbstractElems.ts";
import { treeToString } from "../ImportTree.ts";
import { directive } from "../ParseDirective.ts";
import { parseWgslD } from "../ParseWgslD.ts";
import { testAppParse } from "./TestUtil.ts";
import { assertSnapshot } from "@std/testing/snapshot";

test.ignore("directive parses #export", () => {
  const { appState } = testAppParse(directive, "#export");
  expect(appState[0].kind).toBe("export");
});

test.ignore("parse #export", () => {
  const parsed = parseWgslD("#export");
  expect(parsed[0].kind).toBe("export");
});

test("parse import foo/bar", async (ctx) => {
  const parsed = parseWgslD("import foo/bar");
  await assertSnapshot(ctx, parsed);
});

test("parse module foo.bar.ca", () => {
  const src = `module foo.bar.ca`;
  const appState = parseWgslD(src);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo.bar.ca");
});

test("module foo.bar.ca", (ctx) => {
  const appState = parseWgslD(ctx.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo.bar.ca");
});

test("module foo::bar::ba", (ctx) => {
  const appState = parseWgslD(ctx.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo/bar/ba");
});

test("module foo/bar/ba", (ctx) => {
  const appState = parseWgslD(ctx.name);
  expect(appState[0].kind).toBe("module");
  expect((appState[0] as ModuleElem).name).toBe("foo/bar/ba");
});

test("import ./util/foo;", (ctx) => {
  const appState = parseWgslD(ctx.name);
  const importElem = appState[0] as TreeImportElem;
  const segments = treeToString(importElem.imports);
  expect(segments).toBe("./util/foo");
});

test("import ./bar/foo;", (ctx) => {
  const appState = parseWgslD(ctx.name);
  const importElem = appState[0] as TreeImportElem;
  const segments = treeToString(importElem.imports);
  expect(segments).toBe("./bar/foo");
});

test("import ./bar/{foo,bar};", (ctx) => {
  const appState = parseWgslD(ctx.name);
  const imports = appState.filter((e) => e.kind === "treeImport");
  const segments = imports.map((i) => treeToString(i.imports));
  expect(segments).toContain("./bar/{(foo), (bar)}");
});
