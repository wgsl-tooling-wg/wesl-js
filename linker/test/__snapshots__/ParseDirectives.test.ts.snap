export const snapshot = {};

snapshot[`parse import foo/bar 1`] = `
[
  {
    end: 14,
    imports: ImportTree {
      segments: [
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "foo",
        },
        SimpleSegment {
          args: undefined,
          as: undefined,
          name: "bar",
        },
      ],
    },
    kind: "treeImport",
    start: 0,
  },
]
`;
